common/taskana-common/src/main/java/pro/taskana/common/internal/workingtime/WorkingTimeCalculatorImpl.java

```diff
@@ -0,0 +1,268 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+
+import pro.taskana.common.api.LocalTimeInterval;
+import pro.taskana.common.api.WorkingTimeCalculator;
+
+public class WorkingTimeCalculatorImpl implements WorkingTimeCalculator {
+
+  static final ZoneOffset UTC = ZoneOffset.UTC;
+
+  private final WorkingDaysToDaysConverter converter;
+  private final WorkingTimeSchedule workingTimeSchedule;
+
+  public WorkingTimeCalculatorImpl(
+      WorkingDaysToDaysConverter converter,
+      Map<DayOfWeek, Set<LocalTimeInterval>> workingTimeSchedule) {
+    this.converter = converter;
+    this.workingTimeSchedule = new WorkingTimeSchedule(workingTimeSchedule);
+  }
+
+  @Override
+  public Instant subtractWorkingTime(Instant workStart, Duration workingTime)
+      throws IllegalArgumentException {
+    validatePositiveDuration(workingTime);
+    WorkSlot workSlot = getWorkSlotOrPrevious(toLocalDateTime(workStart));
+    return workSlot.subtractWorkingTime(workStart, workingTime);
+  }
+
+  @Override
+  public Instant addWorkingTime(Instant workStart, Duration workingTime)
+      throws IllegalArgumentException {
+    validatePositiveDuration(workingTime);
+    WorkSlot bestMatchingWorkSlot = getWorkSlotOrNext(toLocalDateTime(workStart));
+    return bestMatchingWorkSlot.addWorkingTime(workStart, workingTime);
+  }
+
+  @Override
+  public Duration workingTimeBetween(Instant from, Instant to) throws IllegalArgumentException {
+    validateProperInterval(from, to);
+
+    WorkSlot bestMatchingWorkSlot = getWorkSlotOrNext(toLocalDateTime(from));
+    Instant earliestWorkStart = max(from, bestMatchingWorkSlot.start);
+    Instant endOfWorkSlot = bestMatchingWorkSlot.end;
+
+    if (endOfWorkSlot.compareTo(to) >= 0) {
+      if (bestMatchingWorkSlot.start.compareTo(to) <= 0) {
+        // easy part. _from_ and _to_ are in the same work slot
+        return Duration.between(earliestWorkStart, to);
+      } else {
+        // _from_ and _to_ are before the bestMatchingWorkSlot aka between two work slots. We simply
+        // drop it
+        return Duration.ZERO;
+      }
+    } else {
+      // Take the current duration and add the working time starting after this work slot.
+      return Duration.between(earliestWorkStart, endOfWorkSlot)
+          .plus(workingTimeBetween(endOfWorkSlot, to));
+    }
+  }
+
+  @Override
+  public boolean isWorkingDay(Instant instant) {
+    return converter.isWorkingDay(toLocalDate(instant));
+  }
+
+  @Override
+  public boolean isWeekend(Instant instant) {
+    return converter.isWeekend(toLocalDate(instant));
+  }
+
+  @Override
+  public boolean isHoliday(Instant instant) {
+    return converter.isHoliday(toLocalDate(instant));
+  }
+
+  @Override
+  public boolean isGermanHoliday(Instant instant) {
+    return converter.isGermanHoliday(toLocalDate(instant));
+  }
+
+  @Override
+  public boolean isWorkingTimeBetween(Instant first, Instant second) {
+    Duration workingTime;
```

Comment:
@mustaphazorgati  I switched back to bringing the parameters in the correct order instead of throwing an exception. At least temporarily. By raising an exception  `acceptance.task.ServiceLevelPriorityAccTest#should_UpdateTaskPlannedOrDue_When_PlannedOrDueAreOnWeekends_ServiceLevel_P0D` failed. That is due to how method `pro.taskana.task.internal.ServiceLevelHandler#ensureServiceLevelIsNotViolated` uses `isWorkingTimeBetween` and for the moment I want to focus on having all tests working before finding a better solution.